### ISSUE TITLE
Pushing to mirror branches

### ISSUE THREAD

**Issue by ghost** [ericpaulbishop/redmine_git_hosting#80]
I have a mirror setup, and I just created a new branch, and pushed it, and it didn't get pushed to the mirror, yet commits in master get pushed fine.

Recap
- Pushing new branch doesn't work
- Deleting a branch probably doesn't work either (goes along with pushing a new branch)
- What about a --force push? If I force a push, does it force to the mirror push as well?

**Comment by ericpaulbishop** [comment]
I just tested this, and it seems to work fine. The --mirror push does force all updates to the designated mirror.

**Comment by ghost** [comment]
Did pushing to new branches (and deleting branches) work too though? It didn't work when I tried.

Mainly, the test I did was, I created a new branch and pushed it, and it didn't get pushed to the mirror.

**Comment by ericpaulbishop** [comment]
Yes, it did work.

**Comment by ghost** [comment]
Alright. I'm going to do some testing on it then.. Maybe there's a bug (or I missed something).

**Comment by ghost** [comment]
Strange, it worked now. Awesome. :) Thanks for all your work, and your patience. :)

**Comment by ghost** [comment]
I just had this problem again. After looking at the code, I realized something wrong is happening.

~git/.ssh/run_gitolite_admin_ssh is being created, but it's being created as:

```

exec ssh -o BatchMode=yes -o StrictHostKeyChecking=no -i /home/git/.ssh/gitolite_admin_id_rsa    "$@"

```

In lib/git_hosting.rb, it says `%x[ #{GitHosting.git_user_runner} 'echo "#!/bin/sh" > ~/.ssh/run_gitolite_admin_ssh']`, but this is not appearing in the generated file!

This is why my mirror push didn't work just now.

Edit: After pushing, the new branch still isn't being created for me, even after I fixed that part. After pushing a new commit, it's not showing up in the branch in Redmine either..

Edit2: After pressing the push button in the settings, it gets pushed correctly. So it seems as if it's not pushing the changes then, even though I have the mirror active and everything.

So, in summary:
- Pushed new branch.
- new branch not pushed to mirror. (push new commit to branch, still not pushed to mirror; also, commit not showing up in Redmine or the database)
- Press push in settings.
- new branch pushed to mirror (previous commit now appears in mirror as well; push new commit to branch, didn't go to mirror again)
- delete branch and push to master
- deleted branch DID get pushed to mirror (and deleted)

This is some odd behavior...

Note: Github is my mirror. So it's possible something weird is happening, like Github not updating quickly. I just manually pushed my new branch to Github, and git said `Everything up-to-date`, and the branch finally showed up, whereas it wasn't before. But this is still really odd. All commits to my main branch work just fine.

**Comment by ghost** [comment]
It really is odd. Because sometimes it works perfectly, but other times (like when I last posted the reply) it didn't work at all, or just partially. I guess it _could_ be GitHub, but I'm still skeptical on that. I just pushed a new branch set to track an upstream branch, and it pushed correctly. The only difference I can think of is that I removed the remote for my GitHub repository (upstream remote is still there), but that shouldn't be related to anything.

The common denominator might be it's pushing to GitHub (still unsure about that though). Is there anyway I can always log the output of my pushes and mirror pushes, so when it happens again, I can give you a log of it?

Edit: Actually, it seems to be as Redmine is the common denominator. Redmine didn't see my commit pushes, even though it saw the branch (which doesn't necessarily mean it fetched the changes correctly). This time it did see the commit push. It could possibly be that the post-receive hook isn't working 100% of the time. I know that there's the debug for the hooks, but I enabled it, and the debug option in the git config for the git user wasn't enabled still.

**Comment by ericpaulbishop** [comment]
Well... what needs to happen to address this is to figure out some set of circumstances where one specific problem can be reliably reproduced.

What concerns me the most of what you have above is that ~git/.ssh/run_gitolite_admin_ssh doesn't have the first line written.  That's.... really, really weird.  Since the code to write the two lines are two lines of ruby code right next to each other, I don't see how this could happen.  My best guess is that there might be a race condition somewhere when the hook ssh script is getting written, otherwise it just makes no sense.

I'll see what I can do, but in order to address the problem we really need to be able to reproduce it reliably.

**Comment by ghost** [comment]
Indeed. I've done some tests of ~git/.ssh/run_gitolite_admin_ssh being generated, and it should be working great.. I remember that it used to work, but not anymore.

Maybe I can place some debugging commands near the part where things get generated, and see what's possibly going on.

Also, #89 might somehow be related to some of the problems here, from the description I read.

**Comment by ghost** [comment]
Updated to latest commit (1233e4ab); I still don't see the `#!/bin/sh` line... Hmm.. However, seeing as your new changes, it might still fix the other weird issues. :)

Edit:: Pushed a new branch, and it worked the first time (beside the ssh file getting in the way because of the missing line). That's a good start, that issue may just be fixed. :)

**Comment by ericpaulbishop** [comment]
I think I figured out what was going on with the first line of the run_gitolite_admin_ssh file, I think the reason I couldn't reproduce it had to do with the different behavior in diffferent versions of sudo (pre 1.7.4 vs 1.7.4+).  Going back to Ubuntu 10.10 allowed me to reproduce the problem.  Should be fixed now.

**Comment by ghost** [comment]
Hi, I verified the problem is now fixed. Thanks a lot! :)
